### PR TITLE
Harmonic oscillation reduction [JSB version]

### PR DESCRIPTION
**Pull request by profjsb**
This is a copy of the PR #285 by @sarajamal57, that has been rebased to the current main of cesium.

It is meant to solve the overflow problem in issue https://github.com/cesium-ml/cesium/issues/284 by:

  - normalization of the input data  if `opt_normalize == True`
  - rescaling of the output dictionary via rescale_lomb_model()

Additional output:
The full power spectrum and associated frequencies are added in the output dictionary ("freqs_vector", "psd_vector")

### CODE REVIEW

File: cesium/features/lomb_scargle.py [hunk 1]
```diff
@@ -84,22 +109,75 @@ def lomb_scargle_model(
...
+        # store current estim_freq results, [**RESCALED for first fund freq]
+        current_fit = (
+            rescale_lomb_model(fit, mmean, mscale)
+            if (opt_normalize & (i == 0))
+            else fit
+        )
+        model_dict["freq_fits"].append(current_fit)
+        model_dict["freq_fits"][-1]["resid"] = norm_residual
+
+        if opt_normalize & (i == 0):
```
Comment (stefanv) [inline cesium/features/lomb_scargle.py:126]: These two if clauses feel as if they should be combined higher up, but perhaps that's tricky.

File: cesium/features/lomb_scargle.py [hunk 2]
```diff
@@ -84,22 +109,75 @@ def lomb_scargle_model(
...
+    Returns
+    -------
+    rescaled_model: dict
+        rescaled Lomb-Scargle model (in adequacy with the input_signal)
+
+    """
+    rescaled_model = norm_model.copy()
+    # unchanged : 'psd', 'chi0', 'freq', 'chi2', 'trace', 'nu0', 'nu', 'npars',
+    #             'rel_phase', 'rel_phase_error', 'time0', 'signif'
+    rescaled_model["s0"] = rescaled_model["s0"] / mscale**2
```
Comment (stefanv) [inline cesium/features/lomb_scargle.py:167]: Maybe three for-loops here?

**Review by stefanv — APPROVED** [review]
Thanks @profjsb. There are some small things that can be tidied up here, but none needs to block merging. Go ahead and do as much as you like, then merge on passing of tests.

File: cesium/features/lomb_scargle.py [hunk 3]
```diff
@@ -84,22 +109,75 @@ def lomb_scargle_model(
...
             )
-        model_dict["freq_fits"].append(fit)
-        signal -= fit["model"]
-        model_dict["freq_fits"][-1]["resid"] = signal.copy()
+
+        # remove current estim_freq for next run [**NORM]
+        norm_residual = signal - fit["model"]
+        signal = norm_residual
+
+        # store current estim_freq results, [**RESCALED for first fund freq]
```
[the remark this answers is not in the document]
Comment (profjsb) [reply cesium/features/lomb_scargle.py:117]: Good idea, done.

File: cesium/features/lomb_scargle.py [hunk 4]
```diff
@@ -71,6 +95,7 @@ def lomb_scargle_model(
                 detrend_order=1,
             )
             model_dict["trend"] = fit["trend_coef"][1]
+        #
```
[the remark this answers is not in the document]
Comment (profjsb) [reply cesium/features/lomb_scargle.py:98]: done

File: cesium/features/lomb_scargle.py [hunk 5]
```diff
@@ -4,11 +4,19 @@
...
 def lomb_scargle_model(
-    time, signal, error, sys_err=0.05, nharm=8, nfreq=3, tone_control=5.0
+    time,
+    signal,
+    error,
+    sys_err=0.05,
+    nharm=8,
+    nfreq=3,
+    tone_control=5.0,
+    opt_normalize=False,
```
[the remark this answers is not in the document]
Comment (profjsb) [reply cesium/features/lomb_scargle.py:14]: good idea. done.

**Pull request merged**